datafusion/src/physical_plan/repartition.rs

```diff
@@ -365,56 +382,98 @@ impl RepartitionExec {
 
         Ok(())
     }
+}
+
+#[derive(Debug)]
+struct AbortOnDrop(Vec<JoinHandle<()>>);
+
+impl Drop for AbortOnDrop {
+    fn drop(&mut self) {
+        for join_handle in &self.0 {
+            join_handle.abort();
+        }
+    }
+}
 
+pin_project! {
     /// Waits for `input_task` which is consuming one of the inputs to
     /// complete. Upon each successful completion, sends a `None` to
     /// each of the output tx channels to signal one of the inputs is
     /// complete. Upon error, propagates the errors to all output tx
     /// channels.
-    async fn wait_for_task(
+    struct WaitForTask {
+        #[pin]
         input_task: JoinHandle<Result<()>>,
         txs: HashMap<usize, UnboundedSender<Option<ArrowResult<RecordBatch>>>>,
-    ) {
+    }
+
+    impl PinnedDrop for WaitForTask {
```

Comment:
As above, I wonder if this could have a single `abort_on_drop` field instead of `input_task` and this `PinnedDrop` implementation

Reply:
done